File: 프로그래머스/lv3/42898. 등굣길/등굣길.py
def solution(m, n, puddles):
    dp = [[-1 for _ in range(m+1)] for _ in range(n+1)] 

    for puddle in puddles:
        dp[puddle[1]][puddle[0]]=0

    check=False
    for i in range(1,n+1):
        if dp[i][1]==0: check=True

        if check: dp[i][1]=0
        else: dp[i][1]=1

    check=False 
    for i in range(1,m+1):
        if dp[1][i]==0: check=True

        if check: dp[1][i]=0
        else: dp[1][i]=1


    for i in range(2,n+1):
        for j in range(2,m+1):
            if dp[i][j]==0: continue
            else:
                dp[i][j]=dp[i-1][j]+dp[i][j-1]

    return dp[-1][-1]%1000000007

File: 프로그래머스/3/42898. 등굣길/등굣길.js
function solution(m, n, puddles) {
    const MOD = 1000000007;

    // dp 배열 초기화
    let dp = Array.from({ length: n + 1 }, () => Array(m + 1).fill(0));

    // 물에 잠긴 지역 처리
    let puddleSet = new Set(puddles.map(([x, y]) => `${y},${x}`));
    console.log(puddleSet)
    // 시작 지점 설정
    dp[1][1] = 1;

    // 동적 계획법을 사용하여 경로 계산
    for (let i = 1; i <= n; i++) {
        for (let j = 1; j <= m; j++) {
            if (i === 1 && j === 1) continue; // 시작점은 이미 처리됨
            if (puddleSet.has(`${i},${j}`)) {
                dp[i][j] = 0; // 물에 잠긴 지역
            } else {
                dp[i][j] = (dp[i - 1][j] + dp[i][j - 1]) % MOD;
            }
        }
    }

    return dp[n][m];
}
